.github/workflows/publish.yml

```diff
@@ -0,0 +1,49 @@
+# This workflow is triggered two ways:
+#
+# 1. When a tag is created, the workflow will upload the package to
+#    test.pypi.org.
+# 2. When a release is made, the workflow will upload the package to pypi.org.
+#
+# It is done this way until PyPI has draft reviews, to allow for a two-stage
+# upload with a chance for manual intervention before the final publication.
+name: Upload package
+
+on:
+  release:
+    types: [created]
+  push:
+    tags:
+      - '*'
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: '3.x'
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install tox
+    - name: Build and publish
+      env:
+        TWINE_USERNAME: "__token__"
+        TWINE_PASSWORD: ${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}
+      run: |
+        if [[ "$GITHUB_EVENT_NAME" == "push" ]]; then
+          export TWINE_REPOSITORY_URL="https://test.pypi.org/legacy/"
+          export TWINE_PASSWORD="${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}"
+        elif [[ "$GITHUB_EVENT_NAME" == "release" ]]; then
+          export TWINE_REPOSITORY_URL="https://upload.pypi.org/legacy/"
+          export TWINE_PASSWORD="${{ secrets.PYPI_UPLOAD_TOKEN }}"
```

Comment:
@pganssle I didn't change the secret names in my suggestion below but I'd like to ask you to rename them to match https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/ (`pypi_password` and `test_pypi_password`) in order to create a more consistent experience across the ecosystem when people discover and try to learn from the workflows implementing publishing to GitHub.

Reply:
I appreciate that, but I have two objections:

1. These are tokens and not passwords – I actually found that confusing in the guide, because you specify passwords and not usernames.
2. I actually think everyone using the same environment variables for this *hurts* peoples' ability to understand what's going on rather than helping it. CI jobs and orchestration and to some extent programming languages are full of "special" variables where the name of the variable itself matters. I think it helps to see people using different variables so that you know that the name is an arbitrary choice and not something baked into GHA.